src/platform/tests/TestMdns.cpp

```diff
@@ -0,0 +1,112 @@
+    nlTestSuite * suite = static_cast<nlTestSuite *>(context);
+
+    NL_TEST_ASSERT(suite, result != nullptr);
+    NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
+    result->mAddress.Value().ToString(addrBuf, sizeof(addrBuf));
+    printf("Service at [%s]:%u\n", addrBuf, result->mPort);
+    NL_TEST_ASSERT(suite, result->mTextEntrySize == 1);
+    NL_TEST_ASSERT(suite, strcmp(result->mTextEntryies[0].mKey, "key") == 0);
+    NL_TEST_ASSERT(suite, strcmp(reinterpret_cast<const char *>(result->mTextEntryies[0].mData), "val") == 0);
+
+    exit(0);
+}
+
+static void HandleBrowse(void * context, MdnsService * services, size_t servicesSize, CHIP_ERROR error)
+{
+    nlTestSuite * suite = static_cast<nlTestSuite *>(context);
+
+    NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
+    if (services)
+    {
+        printf("Mdns service size %zu\n", servicesSize);
+        printf("Service name %s\n", services->mName);
+        printf("Service type %s\n", services->mType);
+        NL_TEST_ASSERT(suite, ChipMdnsResolve(services, INET_NULL_INTERFACEID, HandleResolve, suite) == CHIP_NO_ERROR);
+    }
+}
+
+static void InitCallback(void * context, CHIP_ERROR error)
+{
+    MdnsService service;
+    TextEntry entry;
+    char key[]          = "key";
+    char val[]          = "val";
+    nlTestSuite * suite = static_cast<nlTestSuite *>(context);
+
+    NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
+
+    service.interface = INET_NULL_INTERFACEID;
+    service.mPort     = 80;
+    strcpy(service.mName, "test");
+    strcpy(service.mType, "_mock");
+    service.mProtocol      = MdnsServiceProtocol::kMdnsProtocolTcp;
+    entry.mKey             = key;
+    entry.mData            = reinterpret_cast<const uint8_t *>(val);
+    entry.mDataSize        = strlen(reinterpret_cast<const char *>(entry.mData));
+    service.mTextEntryies  = &entry;
+    service.mTextEntrySize = 1;
+
+    NL_TEST_ASSERT(suite, ChipMdnsPublishService(&service) == CHIP_NO_ERROR);
+    ChipMdnsBrowse("_mock", MdnsServiceProtocol::kMdnsProtocolTcp, INET_NULL_INTERFACEID, HandleBrowse, suite);
+}
+
+static void ErrorCallback(void * context, CHIP_ERROR error)
+{
+    if (error != CHIP_NO_ERROR)
+    {
+        fprintf(stderr, "Mdns error: %d\n", static_cast<int>(error));
+        abort();
+    }
+}
+
+void TestMdnsPubSub(nlTestSuite * inSuite, void * inContext)
+{
+    chip::Platform::MemoryInit();
+    chip::DeviceLayer::PlatformMgr().InitChipStack();
+    NL_TEST_ASSERT(inSuite, chip::Protocols::Mdns::ChipMdnsInit(InitCallback, ErrorCallback, inSuite) == CHIP_NO_ERROR);
+
+    ChipLogProgress(DeviceLayer, "Start EventLoop");
+    chip::DeviceLayer::PlatformMgr().RunEventLoop();
+}
+
+static const nlTest sTests[] = { NL_TEST_DEF("Test Mdns::PubSub", TestMdnsPubSub), NL_TEST_SENTINEL() };
+
+int TestMdns()
+{
+    std::mutex mtx;
+    std::unique_lock<std::mutex> lock(mtx);
+    std::condition_variable done;
+    int retVal = -1;
+
+    std::thread t([&done, &retVal]() {
```

Comment:
Is this thread only for the timeout?

We can't have connect take a timeout & communicate failure back to the test?

Reply:
We can but it's kind of an overkill to add timeout to every mdns api and implement the cancelling mechanism based on the timeout.

Reply:
Does it block forever in production if avahi-daemon isn't running? This seems fairly suspicious to me.

Reply:
All your calls won't block you just doesn't get the success callback.